### PR TITLE
Add another alluxio path replacement algorithm

### PR DESCRIPTION
**Pull request by tgravescs**
fixes https://github.com/NVIDIA/spark-rapids/issues/6500

The original way we replace the paths automatically with the alluxio path causes a lot of overhead to have to list the files again.  That can be a performance hit, especially on very busy clusters.  Here we add another algorithm to replace the paths later that doesn't require a new listing of the files.  This is more performant but doesn't update the metadata like the block locations.  From my investigation currently Alluxio on Databricks is not returning the cached block locations but instead returns the list of all workers, so there is no need to update the locations.  https://github.com/NVIDIA/spark-rapids/issues/6530 to investigate that further.

Open to suggestions on the config value names.

Note that currently we only support the path replacement with parquet datasource v1 so that is all that is change here and we have another issue to track if we change it for other file types and datasources.

Databricks has different versions of FileStatus and PartitionDirectory so had to shim those.

Test this on Databricks on customer cluster and on our own using NDS type queries. Performance if very noticable when using the SELECTION_TIME algorithm when there are a lot of files. We have other issues and plans to improve testing for alluxio.

### CODE REVIEW

**Comment by tgravescs** [comment]
build

**Comment by tgravescs** [comment]
build

File: sql-plugin/src/main/311until320-nondb/scala/com/nvidia/spark/rapids/shims/Spark31XShims.scala [hunk 1]
```diff
@@ -71,6 +71,23 @@ abstract class Spark31XShims extends SparkShims with Spark31Xuntil33XShims with
     fileIndex.allFiles()
   }
 
+  override def alluxioReplacePathsPartitionDirectory(
+      pd: PartitionDirectory,
+      replaceFunc: Option[Path => Path]): (Seq[FileStatus], PartitionDirectory) = {
+    val updatedFileStatus = pd.files.map { f =>
+      val replaced = replaceFunc.get(f.getPath)
+      // Alluxio caches the entire file, so the size should be the same.
+      // Just hardcode block replication to 1 to make sure nothing weird happens but
```
Comment (gerashegalov) [inline sql-plugin/src/main/311until320-nondb/scala/com/nvidia/spark/rapids/shims/Spark31XShims.scala:80]: please clarify "nothing weird", why preserving the replication factor from the original FileStatus is a bad idea

**Comment by tgravescs** [comment]
build

**Comment by tgravescs** [comment]
seems to be something wrong in the jenkins box: 09:35:00  FATAL ERROR in native method: processing of -javaagent failed

**Comment by tgravescs** [comment]
build

**Comment by tgravescs** [comment]
failure is different but again unrelated

**Comment by tgravescs** [comment]
build

**Review by jlowe — APPROVED** [review]

**Pull request merged**